toolchain/lexer/tokenized_buffer_benchmark.cpp

```diff
@@ -0,0 +1,136 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <benchmark/benchmark.h>
+
+#include <forward_list>
+
+#include "absl/random/random.h"
+#include "common/check.h"
+#include "llvm/ADT/Sequence.h"
+#include "llvm/ADT/StringExtras.h"
+#include "toolchain/diagnostics/null_diagnostics.h"
+#include "toolchain/lexer/token_kind.h"
+#include "toolchain/lexer/tokenized_buffer.h"
+
+namespace Carbon::Testing {
+namespace {
+
+class LexerBenchHelper {
+ public:
+  explicit LexerBenchHelper(llvm::StringRef text)
+      : source_(MakeSourceBuffer(text)) {}
+
+  auto Lex() -> TokenizedBuffer {
+    DiagnosticConsumer& consumer = NullDiagnosticConsumer();
+    return TokenizedBuffer::Lex(source_, consumer);
+  }
+
+  auto DiagnoseErrors() -> llvm::StringRef {
+    auto buffer = TokenizedBuffer::Lex(source_, ConsoleDiagnosticConsumer());
+    CARBON_CHECK(buffer.has_errors())
+        << "Asked to diagnose errors but none found!";
+    // Return an empty string so this can be streamed into a check failure.
+    return "";
+  }
+
+ private:
+  llvm::vfs::InMemoryFileSystem fs_;
+  std::string filename_ = "test.carbon";
+  SourceBuffer source_;
+
+  auto MakeSourceBuffer(llvm::StringRef text) -> SourceBuffer {
+    CARBON_CHECK(fs_.addFile(filename_, /*ModificationTime=*/0,
+                             llvm::MemoryBuffer::getMemBuffer(text)));
+    return std::move(*SourceBuffer::CreateFromFile(fs_, filename_));
+  }
+};
+
+constexpr int NumTokens = 100000;
+
+void BM_ValidKeywords(benchmark::State& state) {
+  absl::BitGen gen;
+  std::string source;
+  llvm::raw_string_ostream os(source);
+  llvm::ListSeparator sep(" ");
+  for (int i = 0; i < NumTokens; ++i) {
+    int token = absl::Uniform<int>(gen, 0, TokenKind::KeywordTokens.size());
+    os << sep << TokenKind::KeywordTokens[token].fixed_spelling();
+  }
+
+  LexerBenchHelper helper(source);
+  for (auto _ : state) {
+    TokenizedBuffer buffer = helper.Lex();
+    benchmark::DoNotOptimize(buffer.has_errors());
```

Comment:
Sorry, I'm trying to figure this out, code comments appreciated.

Why is has_errors called, and why the explicit DoNotOptimize? The other path does `CARBON_CHECK(!buffer.has_errors()) << helper.DiagnoseErrors();`, was there an issue with that here? (the check also ensures the parsing isn't faster for some error-related reason)

Alternately, if you only call `helper.Lex()`, was that an issue?

Reply:
Sorry, at one point I was getting failures from `CARBON_CHECK` here and removed it, but they were surrious. Switched back to `CARBON_CHECK`.

A pattern that I find emerges in benchmark code is ensuring that the work done has some consequence so the optimizer doesn't remove it entirely or remove it from the loop breaking the benchmark.

One way of doing that is something like `CARBON_CHECK` that we expect to be very fast when it succeeds and only slow when it fails, but to have a real consequence that keeps code around.

But when I thought that wouldn't work (I was remembering when we lexed keywords that were then rejected; an old state of the lexer), I used `DoNotOptimize` to achieve a similar result for benchmarking without producing an error.